File: src/kodit/bm25/vectorchord_repository.py
"""VectorChord repository for document operations."""

import logging

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from kodit.bm25.keyword_search_service import (
    BM25Document,
    BM25Result,
    KeywordSearchProvider,
)

logger = logging.getLogger(__name__)

# SQL statements
CREATE_VCHORD_EXTENSION = "CREATE EXTENSION IF NOT EXISTS vchord CASCADE;"
CREATE_PG_TOKENIZER = "CREATE EXTENSION IF NOT EXISTS pg_tokenizer CASCADE;"
CREATE_VCHORD_BM25 = "CREATE EXTENSION IF NOT EXISTS vchord_bm25 CASCADE;"
SET_SEARCH_PATH = 'SET search_path TO "$user", public, bm25_catalog, pg_catalog, information_schema, tokenizer_catalog;'

CREATE_BM25_TABLE = """
CREATE TABLE IF NOT EXISTS {table_name} (
    id SERIAL PRIMARY KEY,
    snippet_id BIGINT NOT NULL REFERENCES snippets(id),
    passage TEXT,
    embedding bm25vector,
    UNIQUE(snippet_id)
)
"""

CREATE_BM25_INDEX = """
CREATE INDEX IF NOT EXISTS {index_name}
ON {table_name}
USING bm25 (embedding bm25_ops)
"""

LOAD_TOKENIZER = """
SELECT create_tokenizer('bert', $$
model = "llmlingua2"
pre_tokenizer = "unicode_segmentation"  # split texts according to the Unicode Standard Annex #29
[[character_filters]]
to_lowercase = {}                       # convert all characters to lowercase
[[character_filters]]
unicode_normalization = "nfkd"          # normalize the text to Unicode Normalization Form KD
[[token_filters]]
skip_non_alphanumeric = {}              # skip tokens that all characters are not alphanumeric
[[token_filters]]
stopwords = "nltk_english"              # remove stopwords using the nltk dictionary
[[token_filters]]
stemmer = "english_porter2"             # stem tokens using the English Porter2 stemmer
$$)
"""


class VectorChordRepository(KeywordSearchProvider):
    """Repository for VectorChord document operations."""

    def __init__(
        self,
        session: AsyncSession,
    ) -> None:
        """Initialize the VectorChord repository."""
        self.session = session
        self._initialized = False
        self._bm25_table_name = "vectorchord_bm25_documents"
        self._bm25_index_name = f"{self._bm25_table_name}_idx"

    async def _initialize(self) -> None:
        """Initialize the VectorChord environment."""
        try:
            # Execute each command separately with verification
            await self.session.execute(text(CREATE_VCHORD_EXTENSION))
            await self.session.execute(text(CREATE_PG_TOKENIZER))
            await self.session.execute(text(CREATE_VCHORD_BM25))
            await self.session.execute(text(SET_SEARCH_PATH))
            await self.session.commit()

            # Create the tokenizer
            await self.session.execute(text(LOAD_TOKENIZER))
            await self.session.commit()

            # Create tables with proper dependency order
            await self._create_tables()
            self._initialized = True
        except Exception as e:
            msg = f"Failed to initialize VectorChord repository: {e}"
            raise RuntimeError(msg) from e

    async def _create_tables(self) -> None:
        """Create the necessary tables in the correct order."""
        try:
            await self.session.execute(
                text(CREATE_BM25_TABLE.format(table_name=self._bm25_table_name))
            )
            await self.session.execute(
                text(
                    CREATE_BM25_INDEX.format(
                        table_name=self._bm25_table_name,
                        index_name=self._bm25_index_name,
                    )
                )
            )
            await self.session.commit()
        except Exception as e:
            logger.error(f"Error creating tables: {e}")
            raise

    async def index(self, corpus: list[BM25Document]) -> None:
        """Index a new corpus."""
        if not self._initialized:
            await self._initialize()

        if not corpus:
            return

        # Filter out any documents that don't have a snippet_id or text
        corpus = [
            doc
            for doc in corpus
            if doc.snippet_id is not None and doc.text is not None and doc.text != ""
        ]

        # Write the documents to the bm25 database in one big batch
        query = text(
            f"INSERT INTO {self._bm25_table_name} (snippet_id, passage) VALUES (:snippet_id, :passage)"
        )

        for doc in corpus:
            await self.session.execute(
                query, {"snippet_id": doc.snippet_id, "passage": doc.text}
            )
        await self.session.commit()

        # Tokenize the new documents with schema qualification
        query = text(
            f"UPDATE {self._bm25_table_name} SET embedding = tokenize(passage, 'bert')"
        )
        await self.session.execute(query)
        await self.session.commit()

    async def retrieve(
        self,
        query: str,
        top_k: int = 10,
    ) -> list[BM25Result]:
        """Search documents using BM25 similarity."""
        if not self._initialized:
            await self._initialize()

        if not query or query == "":
            return []

        table_name = self._bm25_table_name
        index_name = self._bm25_index_name

        sql = text(f"""
            SELECT 
                   snippet_id, 
                   embedding <&> to_bm25query('{index_name}', tokenize(:query_text,
                   'bert')) AS bm25_score 
                FROM {table_name} 
                ORDER BY bm25_score
                LIMIT :limit
        """)

        # Bind parameters separately to avoid SQL injection
        params = {"query_text": query, "limit": top_k}

        try:
            result = await self.session.execute(sql, params)
            rows = result.fetchall()

            # Convert to dictionary format
            return [BM25Result(snippet_id=row[0], score=row[1]) for row in rows]
        except Exception as e:
            logger.error(f"Error during BM25 search: {e}")
            raise

File: go-target/internal/indexing/bm25/vectorchord_repository.go
// Package bm25 provides BM25 full-text search implementations.
package bm25

import (
	"context"
	"errors"
	"log/slog"
	"sync"

	"github.com/helixml/kodit/internal/domain"
	"gorm.io/gorm"
)

// SQL statements for VectorChord BM25 operations.
const (
	createVChordExtension = `CREATE EXTENSION IF NOT EXISTS vchord CASCADE`
	createPGTokenizer     = `CREATE EXTENSION IF NOT EXISTS pg_tokenizer CASCADE`
	createVChordBM25      = `CREATE EXTENSION IF NOT EXISTS vchord_bm25 CASCADE`
	setSearchPath         = `SET search_path TO "$user", public, bm25_catalog, pg_catalog, information_schema, tokenizer_catalog`

	createBM25Table = `
CREATE TABLE IF NOT EXISTS vectorchord_bm25_documents (
    id SERIAL PRIMARY KEY,
    snippet_id VARCHAR(255) NOT NULL,
    passage TEXT NOT NULL,
    embedding bm25vector,
    UNIQUE(snippet_id)
)`

	createBM25Index = `
CREATE INDEX IF NOT EXISTS vectorchord_bm25_documents_idx
ON vectorchord_bm25_documents
USING bm25 (embedding bm25_ops)`

	tokenizerExistsQuery = `SELECT 1 FROM tokenizer_catalog.tokenizer WHERE name = 'bert'`

	loadTokenizer = `
SELECT create_tokenizer('bert', $$
model = "llmlingua2"
pre_tokenizer = "unicode_segmentation"
[[character_filters]]
to_lowercase = {}
[[character_filters]]
unicode_normalization = "nfkd"
[[token_filters]]
skip_non_alphanumeric = {}
[[token_filters]]
stopwords = "nltk_english"
[[token_filters]]
stemmer = "english_porter2"
$$)`

	insertQuery = `
INSERT INTO vectorchord_bm25_documents (snippet_id, passage)
VALUES (?, ?)
ON CONFLICT (snippet_id) DO UPDATE
SET passage = EXCLUDED.passage`

	updateEmbeddingsQuery = `
UPDATE vectorchord_bm25_documents
SET embedding = tokenize(passage, 'bert')`

	searchQuery = `
SELECT
    snippet_id,
    embedding <&> to_bm25query('vectorchord_bm25_documents_idx', tokenize(?, 'bert')) AS bm25_score
FROM vectorchord_bm25_documents
ORDER BY bm25_score
LIMIT ?`

	searchQueryWithFilter = `
SELECT
    snippet_id,
    embedding <&> to_bm25query('vectorchord_bm25_documents_idx', tokenize(?, 'bert')) AS bm25_score
FROM vectorchord_bm25_documents
WHERE snippet_id IN ?
ORDER BY bm25_score
LIMIT ?`

	deleteQuery = `DELETE FROM vectorchord_bm25_documents WHERE snippet_id IN ?`

	checkExistingIDsQuery = `SELECT snippet_id FROM vectorchord_bm25_documents WHERE snippet_id IN ?`
)

// ErrInitializationFailed indicates VectorChord initialization failed.
var ErrInitializationFailed = errors.New("failed to initialize VectorChord repository")

// VectorChordRepository implements BM25Repository using VectorChord PostgreSQL extension.
type VectorChordRepository struct {
	db          *gorm.DB
	logger      *slog.Logger
	initialized bool
	mu          sync.Mutex
}

// NewVectorChordRepository creates a new VectorChordRepository.
func NewVectorChordRepository(db *gorm.DB, logger *slog.Logger) *VectorChordRepository {
	if logger == nil {
		logger = slog.Default()
	}
	return &VectorChordRepository{
		db:     db,
		logger: logger,
	}
}

// initialize sets up the VectorChord environment.
func (r *VectorChordRepository) initialize(ctx context.Context) error {
	r.mu.Lock()
	defer r.mu.Unlock()

	if r.initialized {
		return nil
	}

	if err := r.createExtensions(ctx); err != nil {
		return errors.Join(ErrInitializationFailed, err)
	}

	if err := r.createTokenizerIfNotExists(ctx); err != nil {
		return errors.Join(ErrInitializationFailed, err)
	}

	if err := r.createTables(ctx); err != nil {
		return errors.Join(ErrInitializationFailed, err)
	}

	r.initialized = true
	return nil
}

func (r *VectorChordRepository) createExtensions(ctx context.Context) error {
	db := r.db.WithContext(ctx)

	if err := db.Exec(createVChordExtension).Error; err != nil {
		return err
	}
	if err := db.Exec(createPGTokenizer).Error; err != nil {
		return err
	}
	if err := db.Exec(createVChordBM25).Error; err != nil {
		return err
	}
	if err := db.Exec(setSearchPath).Error; err != nil {
		return err
	}
	return nil
}

func (r *VectorChordRepository) createTokenizerIfNotExists(ctx context.Context) error {
	db := r.db.WithContext(ctx)

	var exists int
	result := db.Raw(tokenizerExistsQuery).Scan(&exists)
	if result.Error != nil && !errors.Is(result.Error, gorm.ErrRecordNotFound) {
		return result.Error
	}

	// Tokenizer doesn't exist if RowsAffected is 0
	if result.RowsAffected == 0 {
		if err := db.Exec(loadTokenizer).Error; err != nil {
			return err
		}
	}

	return nil
}

func (r *VectorChordRepository) createTables(ctx context.Context) error {
	db := r.db.WithContext(ctx)

	if err := db.Exec(createBM25Table).Error; err != nil {
		return err
	}
	if err := db.Exec(createBM25Index).Error; err != nil {
		return err
	}
	return nil
}

func (r *VectorChordRepository) existingIDs(ctx context.Context, ids []string) (map[string]struct{}, error) {
	if len(ids) == 0 {
		return map[string]struct{}{}, nil
	}

	var existingIDs []string
	err := r.db.WithContext(ctx).Raw(checkExistingIDsQuery, ids).Scan(&existingIDs).Error
	if err != nil {
		return nil, err
	}

	result := make(map[string]struct{}, len(existingIDs))
	for _, id := range existingIDs {
		result[id] = struct{}{}
	}
	return result, nil
}

// Index adds documents to the BM25 index.
func (r *VectorChordRepository) Index(ctx context.Context, request domain.IndexRequest) error {
	if err := r.initialize(ctx); err != nil {
		return err
	}

	documents := request.Documents()

	// Filter out invalid documents
	var valid []domain.Document
	for _, doc := range documents {
		if doc.SnippetID() != "" && doc.Text() != "" {
			valid = append(valid, doc)
		}
	}

	if len(valid) == 0 {
		r.logger.Warn("corpus is empty, skipping bm25 index")
		return nil
	}

	// Filter out already indexed documents
	ids := make([]string, len(valid))
	for i, doc := range valid {
		ids[i] = doc.SnippetID()
	}

	existing, err := r.existingIDs(ctx, ids)
	if err != nil {
		return err
	}

	var toIndex []domain.Document
	for _, doc := range valid {
		if _, exists := existing[doc.SnippetID()]; !exists {
			toIndex = append(toIndex, doc)
		}
	}

	if len(toIndex) == 0 {
		r.logger.Info("no new documents to index")
		return nil
	}

	// Execute inserts in a transaction
	return r.db.WithContext(ctx).Transaction(func(tx *gorm.DB) error {
		for _, doc := range toIndex {
			if err := tx.Exec(insertQuery, doc.SnippetID(), doc.Text()).Error; err != nil {
				return err
			}
		}

		// Tokenize the new documents
		if err := tx.Exec(updateEmbeddingsQuery).Error; err != nil {
			return err
		}

		return nil
	})
}

// Search performs BM25 keyword search.
func (r *VectorChordRepository) Search(ctx context.Context, request domain.SearchRequest) ([]domain.SearchResult, error) {
	if err := r.initialize(ctx); err != nil {
		return nil, err
	}

	query := request.Query()
	if query == "" {
		return []domain.SearchResult{}, nil
	}

	topK := request.TopK()
	if topK <= 0 {
		topK = 10
	}

	var rows []struct {
		SnippetID  string  `gorm:"column:snippet_id"`
		BM25Score  float64 `gorm:"column:bm25_score"`
	}

	var err error
	snippetIDs := request.SnippetIDs()
	if len(snippetIDs) > 0 {
		err = r.db.WithContext(ctx).Raw(searchQueryWithFilter, query, snippetIDs, topK).Scan(&rows).Error
	} else {
		err = r.db.WithContext(ctx).Raw(searchQuery, query, topK).Scan(&rows).Error
	}

	if err != nil {
		return nil, err
	}

	results := make([]domain.SearchResult, len(rows))
	for i, row := range rows {
		// VectorChord returns negative scores (higher is better when more negative)
		// Convert to positive scores for consistency
		results[i] = domain.NewSearchResult(row.SnippetID, -row.BM25Score)
	}

	return results, nil
}

// Delete removes documents from the BM25 index.
func (r *VectorChordRepository) Delete(ctx context.Context, request domain.DeleteRequest) error {
	if err := r.initialize(ctx); err != nil {
		return err
	}

	ids := request.SnippetIDs()
	if len(ids) == 0 {
		return nil
	}

	return r.db.WithContext(ctx).Exec(deleteQuery, ids).Error
}
